### DISCUSSION TITLE
Controlling solenoid with PWM after ON period

### DISCUSSION THREAD

**Discussion by Mario-75** [arendst/Tasmota#21584] [Q&A]
I’m controlling a solenoid with my WemosD1mini. The solenoid gets full power (ON) for about 50ms to open completely. After that time the power is lowered by using PWM signal (e.g. 50%) to reduce heat and save energy. In detail, every solenoid has a defined ON time and a level of PWM where it stays ON. 
I’m planning to use that device with Tasmota. Unfortunately, I cannot find any information if that starting sequence is possible. I’ve read the PWM section but this seems to be more for dimming lights.
Does anyone know if that is possible?
Thanks

**Comment by sfromis** [comment]
If you configure for basic PWM `SetOption15` (instead of the default as a light), you can select PWM duty cycle using the PWM command. If you issue multiple commands in a sequence, there is a default of 200 ms between commands, which you can tweak using `SetOption34`. I think you could do something like:
`Backlog PWM1 1023; PWM1 511`
Note that the processing time may mean that the actual period between the two commands is longer than the inter-command delay you config.

**Reply by Mario-75** [reply]
Thanks for the answer. I will try your suggestions as soon as I get one of my WeMos Mini to work. 
For unknown reason they keep resing themself. I tried different Wemos (12F and Pro) different Tasmota version 12, 13 and 14 but all reset. I'm using VS Code with Platformio.

**Reply by sfromis** [reply]
I'd recommend starting with one of the precompiled binaries, and better just add one config at a time, depending on what details are behind "keep resetting".

**Comment by Mario-75** [comment]
I had a button connected to D8 GPIO15. That was causing problems for the Wemos.
I further developed my idea of controlling the solenoid by using scripting language. I can now "control" the full power ON period more or less accurate, but good enough for me.
My script isn’t finish yet and I hope it will fit into the script buffer!
I have lots of ; and prints for better understanding what’s going on

```
>D
tmp=0
pub1=0		;mqtt send flag

v1=0		;Valve1 
ut1=0		;use Timer
deft1=30	;default t1 timer
txt1=0		;current value of timer
fpo=1023 	;full power ON
mpo1=500	;PWM reduced ON
fpot1=1000	;full power ON Time to be visable
t:timer1=-1	;timer for automatic OFF

;Second Valve later
sw2=0
ut2=0
txt2=36

>B
pwm1 off
print +++B After Boot
print +++B section v1:%0v1%
print +++B section v1:%0v1%
print +++B section button1:%0bt1[1]%
print +++B section Power1:%0pwr[1]%
txt1=deft1
print +++B section Default Timerzeit: %0txt1%
print +++B section Default Timerzeit: %deft1%

>b
print +++ in b Section Button Pressed 
if bt[1]>0
then
v1^=1
print +++ Switch toggle Valve1 state: %0v1%
endif

>F

if chg[v1]>0{
	if v1>0{
		print +++In F section Power ON v1= %0v1%
		->pwm1 %fpo%
		ts1(fpot1)
		=>publish stat/%topic%/Valve1 {"Valve":%0v1%, "timer":%0ut1%,"countdown":%0txt1%,"Settings": {PWMLevel":%0mpo1%, "fullOnTime":%0fpot1%, "setTimeOut":%0deft1%}}
		if ut1==1{
			print af use timer, start countdown
			timer1=deft1
		}	
	}else{
		print +++In F section Power turns OFF  v1= %0v1%
		->pwm1 off
		timer1=-1
		=>publish stat/%topic%/Valve1 {"Valve":%0v1%, "timer":%0ut1%,"countdown":%0txt1%,"Settings": {PWMLevel":%0mpo1%, "fullOnTime":%0fpot1%, "setTimeOut":%0deft1%}}
		print +++In F section var deft1 = %deft1%
		txt1=deft1	
		print +++In F section var deft1 = %deft1%	
	}	
}

>E

>P
-print in P Section

>S
; stop timer after expired
if timer1==0
then 
txt1=0
timer1=-1
print +++ timer1 expired Switch1: %0cw1%, valve: %0v1%
v1=0
endif
if upsecs%10==0{
	if timer1>0{
		txt1=timer1
		if mqtts>0{
			=>publish stat/%topic%/Valve1 {"countdown":%0txt1%}
		}
	}
}
;send full MQTT struct once MQTT is connected
if pub1==0
and mqtts>0
then
=#publish
pub1=1
endif

>ti1
->pwm1 %mpo1%
print +++ PWM full Power timer end reached

>W
<style>table tbody tr th{color:#ffffff;}table tbody tr td{color:#ffffff;}h2{color:#fff;}h4{color:#f1b24c;}#ptd{color:#f3e463;padding:0px;}</style>
so(4)
<h4>Valve 1</h4>
bu(v1 "Valve1 is on" "Valve1 is off")
ck(ut1 "Use Timer1")
%=#wsub1
<h4>Valve 2</h4>
bu(sw2 "Valve2 is on" "Valve2 is off")
ck(ut2 "Use Timer2")
%=#wsub2

; display web UI
#wsub1
if ut1==1
then
wcs <p style="text-align:left;">Countdown: <span style="float:right;">%2.0(int(txt1/60))%:%2.0(txt1%60)% min</span></p>
=>publish stat/%topic%/Valve1 {"timer":%0ut1%}
endif

#wsub2
if ut2==1
then
wcs <p style="text-align:left;">Countdown: <span style="float:right;">%2.0(int(txt2/60))%:%2.0(txt2%60)% min</span></p>
endif

#publish
->print Publisch MQTT construct
=>publish stat/%topic%/Valve1 {"Valve":%0v1%, "timer":%0ut1%,"countdown":%0txt1%,"Settings": {PWMLevel":%0mpo1%, "fullOnTime":%0fpot1%, "setTimeOut":%0deft1%}}
+>subscribe v1, cmnd/%topic%/Valve1/Valve
+>subscribe ut1, cmnd/%topic%/Valve1/timer
+>subscribe deft1, cmnd/%topic%/Valve1/Settings/setTimeOut
```
